File: paper-ai-reader/frontend/src/types/annotation.ts
export interface AnnotationRect {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface AnnotationPosition {
  page: number;
  rects: AnnotationRect[];
  scale: number;
}

export interface Annotation {
  id: number;
  paper_id: number;
  page_number: number;
  selected_text?: string | null;
  note?: string | null;
  color: string;
  position_json?: string | null;
  annotation_type: string;
  created_at: string;
  updated_at: string;
}

export interface AnnotationCreate {
  page_number: number;
  selected_text?: string | null;
  note?: string | null;
  color?: string;
  position_json?: string | null;
  annotation_type?: string;
}

export interface AnnotationUpdate {
  page_number?: number;
  selected_text?: string | null;
  note?: string | null;
  color?: string;
  position_json?: string | null;
  annotation_type?: string;
}

File: paper-ai-reader/backend/app/schemas/annotation.py
from datetime import datetime

from pydantic import ConfigDict
from sqlmodel import SQLModel


class AnnotationBase(SQLModel):
    page_number: int
    selected_text: str | None = None
    note: str | None = None
    color: str = "#facc15"
    position_json: str | None = None
    annotation_type: str = "highlight"


class AnnotationCreate(AnnotationBase):
    paper_id: int


class PaperAnnotationCreate(AnnotationBase):
    pass


class AnnotationUpdate(SQLModel):
    page_number: int | None = None
    selected_text: str | None = None
    note: str | None = None
    color: str | None = None
    position_json: str | None = None
    annotation_type: str | None = None


class AnnotationRead(AnnotationBase):
    id: int
    paper_id: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)
